### ISSUE TITLE
[SYCL][CUDA] redefinition of 'KernelInfo'

### ISSUE THREAD

**Issue by zjin-lcf** [intel/llvm#4472]
Building an example https://github.com/zjin-lcf/oneAPI-DirectProgramming/tree/master/babelstream-sycl with the lastestSYCL compiler with CUDA support shows the following errors. However, using an older version does not produce these errors. Thanks for the instructions.

```
clang++ -DUSE_GPU -std=c++17 -Wall -I../include -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -Xsycl-target-backend "--cuda-gpu-arch=sm_xx" --gcc-toolchain="/path/to/gcc" -O3 -c main.cpp -o main.o

clang-14: warning: unknown CUDA version: cuda.h: CUDA_VERSION=11040.; assuming the latest supported version 10.1 [-Wunknown-cuda-version]
In file included from <built-in>:1:
/tmp/main-header-1214d7.h:240:20: error: redefinition of 'KernelInfo<init_kernel>'
template <> struct KernelInfo<init_kernel> {
                   ^~~~~~~~~~~~~~~~~~~~~~~
/tmp/main-header-1214d7.h:128:20: note: previous definition is here
template <> struct KernelInfo<init_kernel> {
                   ^
/tmp/main-header-1214d7.h:256:20: error: redefinition of 'KernelInfo<copy_kernel>'
template <> struct KernelInfo<copy_kernel> {

```

**Comment by Michoumichmich** [comment]
Hello @zjin-lcf, shouldn't your kernel names take the template parameter `T` to avoid redefinition ? Furthermore I think that they should be forward declared. So shouldn't you have somewhere 
```
template <typename T>
class copy_kernel;
``` 
?

**Comment by zjin-lcf** [comment]
I will follow 4.12.2. The examples there are helpful.  Thank you for the suggestion! 
I am not clear if '-fsycl-unnamed-lambda' could hide the issues you mentioned.

**Comment by npmiller** [comment]
@zjin-lcf do you think there is something we should look at in DPC++ or is this resolved? It looks like the behavior changed but it seems to me that the new behavior is what is mandated by the SYCL spec, I'm not sure why it worked before.

**Comment by zjin-lcf** [comment]
When a kernel name is specified, then the compiler will check the correctness of the name in any case. Is this right ?
Thanks

**Comment by npmiller** [comment]
> When a kernel name is specified, then the compiler will check the correctness of the name in any case. Is this right ?
> Thanks

As far as I can tell it should check in any case yes, the original sample also fails with `-fsycl-unnamed-lambda` as that shouldn't affect named kernels.
